sympy/sets/sets.py

```diff
@@ -534,6 +535,60 @@ def measure(self):
         """
         return self._measure
 
+    @property
+    def kind(self):
+        """
+        The kind of a Set.
+
+        Every instance of Set or its subclass will have kind of class
+        ``SetKind``.
+        Parameter will be kind of elements of the sets
+
+        Parameter of SetKind can be ``NumberKind``, ``UndefinedKind``,
+        ``TupleKind`` etc.
+        if the elements of a set do not all have the same kind then the kind
+        will be ``SetKind(UndefinedKind)``
+
+        Examples
+        ========
+
+        >>> from sympy import Interval, Matrix, FiniteSet, EmptySet, ProductSet, PowerSet
+
+        >>> FiniteSet(Matrix([1, 2])).kind
+        SetKind(MatrixKind(NumberKind))
+
+        kinds of 0 and Matrix are different, kind will be
+        SetKind(UndefinedKind):
+
+        >>> FiniteSet(0, Matrix([1, 2])).kind
+        SetKind(UndefinedKind)
```

Comment:
Maybe say:

When the elements of the set do not all have the same kind the kind will be returned as SetKind(UndefinedKind). This example should come last in the series of examples.